### PR TITLE
RDRP-560: NI inclusion in TAU, and NI SAS created

### PR DESCRIPTION
**Pull request by JenCheshire**
Created `config/output_schemas/ni_sas_schema.toml` (see notes below re: commented out items)
Run 'run_ni_headcount_fte()' in `src/northern_ireland/ni_main.py`
Moved the output preparation functions to `src/outputs/form_output_prep.py` an added required code for NI prep (where NI data available)
Created `src/outputs/ni_sas.py` for NI SAS output
Add NI SIC mapping to `run_pg_conversion()`
Add NI data to TAU outputs

Couple of things to note at the moment:
•	There are some columns commented out of the NI schema – we are just waiting for confirmation from team on NI schema
•	NI data has period_year, rather than period – column name changed so it pulls to all outputs. Note, NI is YYYY rather than YYYYMM
•	All changes have been made – including where we’ve had updates in the chat and the additional ticket 631.  
•	I’ve put the PG mapping in output prep at the moment – if I put them in staging, pg_numeric and formtype are being dropped after construction.  This is easily moved if the construction schema is changed.
•	I have done the full mapping process on the NI data as implied in the chat.  This can be changed if they don’t want the gaps filled.  The gaps in the alphapg (rtngrpno) mirror the gaps we already know about and are being looked at it another ticket.

Closes RDRP-560, RDRP-426, RDRP-631

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: config/output_schemas/ni_sas_schema.toml [hunk 1]
```diff
@@ -0,0 +1,151 @@
+[ref]
```
Comment (zorge69) [inline config/output_schemas/ni_sas_schema.toml:1]: NI SAS schema is exactly the same as GB SAS schema. Shall we keep just one toml file, "sas_schema.toml" and repeat it two times in developer config:

gb_sas_schema: "config/output_schemas/sas_schema.toml"
ni_sas_schema: "config/output_schemas/sas_schema.toml"

File: src/northern_ireland/ni_staging.py
Comment (zorge69) [inline src/northern_ireland/ni_staging.py:1]: Line 9: run_ni_headcount_fte is imported but not used in ni_staging.py

File: config/output_schemas/ni_sas_schema.toml (hunk 1 above)
Comment (JenCheshire) [reply config/output_schemas/ni_sas_schema.toml:1]: Yes - good idea if the fields are going to be the same.  We are waiting for confirmation if all cols are required, as some cols are missing.  I have commented these out in this version.   When we have had the final schema from the team, can we review then?

File: src/northern_ireland/ni_main.py
Comment (zorge69) [inline src/northern_ireland/ni_main.py:1]: Docstring in line 20 says
    """Run the Northern Ireland module to stage and apply construction.
It probably should say to stage NI data and apply construction to it

**Review by AnneONS — APPROVED** [review]
Looks great

File: src/northern_ireland/ni_main.py
Comment (JenCheshire) [reply src/northern_ireland/ni_main.py:1]: Changed

File: src/northern_ireland/ni_staging.py
Comment (JenCheshire) [reply src/northern_ireland/ni_staging.py:1]: Removed this and os import not used

File: src/outputs/outputs_main.py [hunk 2]
```diff
@@ -142,6 +147,19 @@ def run_outputs(
...
 
+    # Running NI SAS output
+    if config["global"]["output_ni_sas"]:
+        OutputMainLogger.info("Starting NI SAS output...")
+        output_ni_sas(
+            ni_full_responses,
+            config,
+            write_csv,
+            run_id,
+            ultfoc_mapper,
```
Comment (zorge69) [inline src/outputs/outputs_main.py]: We have a column "foc". We don't need a mapper. Just rename "foc" into "f_owner" and make sure the column appear in correct order.

**Pull request merged**